src/arduino/Adafruit_USBD_CDC.cpp

```diff
@@ -33,18 +33,14 @@
 #include "Adafruit_USBD_CDC.h"
 #include "Adafruit_USBD_Device.h"
 
-// TODO Multiple instances supports
-//	static uint8_t _itf_count;
-//	static Adafruit_USBD_CDC* _itf_arr[]
-
-#define EPOUT 0x00
+#define EPOUT 0x00   // Starting endpoints; adjusted elsewhere as needed
 #define EPIN 0x80
 
 Adafruit_USBD_CDC Serial;
 
-Adafruit_USBD_CDC::Adafruit_USBD_CDC(void) {
+Adafruit_USBD_CDC::Adafruit_USBD_CDC(int instance) {
```

Comment:
the _instance should be assigned based on the order of cdc.begin(), this make sure it is consistent with the dynamic configuration generation. I would suggest to have an static variable of the classthat increase each begin.

Reply:
Yeah, but the instances actually returned in the callbacks are created and ordered much later - when the host is requesting info on the "interfaces."  As far as I could figure out:

1. `Adafruit_USBD_CDC foo;` creates the data structures used by the Arduino-specific code.
2. `foo.begin(...)` adds the CDC info to a TinyUSB dynamic configuration list.
3. The host says "hi.  Tell me about yourself"
4. TinyUSB parses the first part of its config list and says "I'm a composite device with 4 interfaces."
5. The host says "tell me about interface #1"
6. TinyUSB parses more of its config list and says:  "It's a CDC/ACM.  blah blah."  It's at this point that it fills in the "next available" slot in `_itf_arr[]`, and the "next available" becomes the index ("Instance") that is used in the callbacks, but I'm not sure that this index is related in any way to anything provided earlier, or even the order of the `foo.begin()` calls (does the host actually have to ask for interface configurations "in-order"?)

I guess that we could strongly-link the index to the creation order, but it seems like it would need to be more invasive of the existing code; probably adding a "desired instance #" to one or more of the data structures, and replacing the "find free _itf_arr[] entry" with a "find matching", but I wanted to change as little core code as possible.

(We could continue this conversation in the TinyUSB project "discussion" that I started a little while ago; it might be a better spot.  But I figured a pull request was a good way to get a lot of exposition out of the way.)

Reply:
I don't quite understand your point, can you sum it up. 

There is issue if one of the cdc instance does not call begin() or they don't call begin() by the order of creation. 
Also there is scenario when user want dynamic configuration where it will appear as single port or dual cdc port (along with other class combination) depending on its own logic.

> (We could continue this conversation in the TinyUSB project "discussion" that I started a little while ago; it might be a better spot. But I figured a pull request was a good way to get a lot of exposition out of the way.)

This is not really relevant to the tinyusb stack, it will always count the instance starting from 0 to n in the order it found in the configuration descriptor.  That is why I think

Reply:
Summary: TinyUSB assigns "instance" numbers class/cdc/cdc_device.c:cdcd_open(), called from process_control_request(), in the order in which the control requests come in for each interface.  I interpreted that the requests come from the host, and I'm not certain whether they are in the same order that the begin() calls happened.  So there's no forward or backward connection between the begin() calls from the sketch and the instance numbers in the callbacks, and I don't know how to create one (or whether it's necessary.)

Reply:
I see, the instance number is still not all clear to you. I will do the modification for this one and push update when I have time.

Reply:
Oh wait - did you mean that I could simply create the instance values in the Adafruit_USBD_CDC::begin() code (incrementing the value as the begin()s are done) - that would be pretty easy.  I was worried about the existing CDC example that demonstrates the ability of CDCs to be turned off via Serial.end()

Reply:
Yeah, that is exactly what I mean, I have updated the PRs and also add guard for cdc usage along with other minor works. Though I couldn't push to your fork to update the PR. Normally write-contributors also gain write access to fork repo to help with PR. Please check out your fork setting, to make sure I could push update to it.